tpu-client-next/src/workers_cache.rs

```diff
@@ -279,3 +311,149 @@ pub fn shutdown_worker(worker: ShutdownWorker) {
         }
     });
 }
+
+#[cfg(test)]
+mod tests {
+    use {
+        crate::{
+            connection_workers_scheduler::BindTarget,
+            quic_networking::{create_client_config, create_client_endpoint},
+            send_transaction_stats::SendTransactionStatsNonAtomic,
+            transaction_batch::TransactionBatch,
+            workers_cache::{spawn_worker, WorkersCache, WorkersCacheError},
+            SendTransactionStats,
+        },
+        solana_tls_utils::QuicClientCertificate,
+        std::{
+            net::{Ipv4Addr, SocketAddr},
+            sync::Arc,
+            time::Duration,
+        },
+        tokio::time::{sleep, timeout, Instant},
+        tokio_util::sync::CancellationToken,
+    };
+
+    // Specify the pessimistic time to finish generation and result checks.
+    const TEST_MAX_TIME: Duration = Duration::from_secs(5);
+
+    #[tokio::test]
+    async fn test_worker_stopped_after_failed_connect() {
+        let address = SocketAddr::new(Ipv4Addr::new(127, 0, 0, 1).into(), 0);
```

Comment:
please use `localhost_port_range_for_tests` when binding to localhost. It gives you non-overlapping port ranges that prevent conflicts between tests.